#5 dump data service

bion33 opened the pull request.

bion33 committed:
Implement Dump Parsing (#5)

Needs testing.

bion33 committed:
Undo accidental unrelated change

drehtisch commented on NationStatesAPIBot/NationStatesAPIBot/DumpData/CAUSE.cs:
```diff
@@ -0,0 +1,10 @@
+﻿namespace DumpData
```
Could you use something like 'NationStatesAPIBot.DumpData' as Namespace instead?

bion33 replied:
Sure, I see now that that's more consistent with the rest

drehtisch commented on NationStatesAPIBot/NationStatesAPIBot/Services/DumpDataService.cs:
```diff
@@ -1,6 +1,245 @@
+
+namespace NationStatesAPIBot.Services
 {
     class DumpDataService
```
Please make it public or at least Internal.

drehtisch commented on NationStatesAPIBot/NationStatesAPIBot/Services/DumpDataService.cs:
```diff
@@ -1,6 +1,245 @@
+        private static List<REGION> _regions;
+        private static Task _updating;
+
+        public DumpDataService(ILogger<RecruitmentService> logger, IOptions<AppSettings> appSettings, BaseApiService apiService, CancellationToken cancel)
```
You aren't using appSettings and logger are you?

bion33 replied:
I'll remove the unused parameters.

drehtisch commented on NationStatesAPIBot/NationStatesAPIBot/Services/DumpDataService.cs:
```diff
@@ -1,6 +1,245 @@
+
+        public Task Update()
+        {
+            return _updating = Task.Run(async () => {
```
Using await Task.Run isn't necessary.

bion33 replied:
If I don't use Task.Run, how could I best wait for the completion of _updating in GetNationsAsync & GetRegionsAsync?

drehtisch replied:
Just by e.g. await GetNationsAsync() it will return when the call is done. I can send you ressources about that if want to. Just let me know.

bion33 replied:
Those will probably come in handy. I don't understand how to wait for a Task without a reference to a Task object.

drehtisch replied:
https://docs.microsoft.com/en-us/dotnet/csharp/async
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/index

drehtisch commented on NationStatesAPIBot/NationStatesAPIBot/Services/DumpDataService.cs:
```diff
@@ -1,6 +1,245 @@
+        
+        private static List<REGION> GetRegionsFromCompressedStream(GZipStream stream)
+        {
+            Console.WriteLine("Extracting compressed stream to REGION Collection");
```
Please use logger instead of Console.WriteLine

bion33 replied:
Totally overlooked those xD

drehtisch replied:
When logging please use LogEventId Provider and LogMessageBuilder see BasicNationStats for reference.

drehtisch reviewed (changes requested):
Looks good so far. Just commented a few minor things.

bion33 committed:
Changes after review (#5)

(Except removing Task.Run, waiting for discussion)

bion33 committed:
Public methods to get NATION(S) and REGION(S)

bion33 committed:
Removed Task.Run, added method to wait for Update

drehtisch commented:
Please have a look on my improvement suggestions #21

drehtisch committed:
Improve 5# suggestion (#21)

* Some Preparerations

* Some Additional Logging

- DataUnavailableException

* DumpDataService Improvements

- Remove Static
- Periodical Update -> UpdateCycle
- Caching of dumps to local file system
- Intial Update -> load cached dumps if available
- More logging
- Use WaitForAction for Waiting
- WaitForDataAvailability

* Add IsReadyForDownloadOfDumps

* Add Equals and GetHashCode

* Start DumpUpdateCycle && Fix return of closed stream.

* Change Nation Collection to HashSet

* Stop DumpUpdateCycle and StopRecruitment

* Performance Boost

* PerformanceBoost 2 - change iteration order

* Precaution for timing issues, Performance Boost 3

* Fix Writing to local fs.

drehtisch committed:
Merge branch 'master' into #5_Dump_Data_Service

drehtisch reviewed (approved):
👍
